AmazonLambdaRecorder Handler Discovery Erroneously Considers Decorators

### Describe the bug

In the absence of any configuration, the `io.quarkus.amazon.lambda.runtime.AmazonLambdaRecorder` automatically discovers the application's single `RequestHandler` or `RequestStreamHandler` class. If multiple candidate types are discovered this way, then the `AmazonLambdaRecorder` errors:

```
java.lang.RuntimeException: Multiple handler classes, either specify the quarkus.lambda.handler property, or make sure there is only a single com.amazonaws.services.lambda.runtime.RequestHandler or, com.amazonaws.services.lambda.runtime.RequestStreamHandler implementation in the deployment
```

Presently, this type discovery logic considers `RequestHandler` and `RequestStreamHandler` types that are CDI decorators (i.e. annotated with `jakarta.decorator.Decorator`). This doesn't seem right, because the decorator type itself can never be the request handler, absent the handler it decorates.

This behavior is a problem for application developers that want to seamlessly introduce new decorators for their Lambda function handlers, because the introduction of the decorator unexpectedly requires the application develop to name all lambda function handler type beans and introduce the `quarkus.lambda.handler` configuration property.

### Expected behavior

`AmazonLambdaRecorder` ignores CDI decorator types in its `RequestHandler` and `RequestStreamHandler` type discovery. Therefore, it selects the first not-decorator `RequestHandler` or `RequestStreamHandler` bean, and, of course, that is decorated with any applicable `jakarta.decorator.Decorator` beans that may exist.

### Actual behavior

Quarkus reports a conflict between the `RequestHandler` or `RequestStreamHandler` and the decorator, and this conflict must be resolved using `quarkus.lambda.handler` to specify the not-decorator type.

### How to Reproduce?

[lambda-request-handler-decorator.zip](https://github.com/quarkusio/quarkus/files/12083273/lambda-request-handler-decorator.zip)


### Output of `uname -a` or `ver`

Darwin pixee-mbp-gilday.localdomain 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18 OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7) OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.2.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.8 (4c87b05d9aedce574290d1acc98575ed5eb6cd39) Maven home: /Users/jgilday/.m2/wrapper/dists/apache-maven-3.8.8-bin/67c30f74/apache-maven-3.8.8 Java version: 17.0.7, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home Default locale: en_US, platform encoding: UTF-8 OS name: "mac os x", version: "13.4.1", arch: "aarch64", family: "mac"

### Additional information

_No response_

This is also an issue at build time for Lambda functions that use the AWS Lambda HTTP extension: decorator beans on the classpath that would otherwise do nothing cause build failures.

```
        Caused by: io.quarkus.builder.BuildException: Build failure: Multiple handler classes.  You have a custom handler class and the AWS Lambda HTTP extension.  Please remove one of them from your deployment.
                at io.quarkus.amazon.lambda.deployment.AmazonLambdaProcessor.discover(AmazonLambdaProcessor.java:88)

```

--- a/extensions/amazon-lambda/deployment/pom.xml
+++ b/extensions/amazon-lambda/deployment/pom.xml
@@ -29,6 +29,11 @@
             <artifactId>rest-assured</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
 
         <dependency>
             <groupId>io.quarkus</groupId>

--- a/extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/AmazonLambdaProcessor.java
+++ b/extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/AmazonLambdaProcessor.java
@@ -3,13 +3,13 @@ package io.quarkus.amazon.lambda.deployment;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.Predicate;
 
 import jakarta.inject.Named;
 
@@ -29,6 +29,7 @@ import io.quarkus.amazon.lambda.runtime.FunctionError;
 import io.quarkus.amazon.lambda.runtime.LambdaBuildTimeConfig;
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
+import io.quarkus.arc.processor.DotNames;
 import io.quarkus.builder.BuildException;
 import io.quarkus.deployment.Feature;
 import io.quarkus.deployment.annotations.BuildProducer;
@@ -58,6 +59,18 @@ public final class AmazonLambdaProcessor {
     private static final DotName NAMED = DotName.createSimple(Named.class.getName());
     private static final Logger log = Logger.getLogger(AmazonLambdaProcessor.class);
 
+    private static final Predicate<ClassInfo> INCLUDE_HANDLER_PREDICATE = new Predicate<>() {
+
+        @Override
+        public boolean test(ClassInfo classInfo) {
+            if (classInfo.isAbstract() || classInfo.hasAnnotation(DotNames.DECORATOR)) {
+                return false;
+            }
+
+            return true;
+        }
+    };
+
     @BuildStep
     FeatureBuildItem feature() {
         return new FeatureBuildItem(Feature.AMAZON_LAMBDA);
@@ -75,11 +88,13 @@ public final class AmazonLambdaProcessor {
             BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchy,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClassBuildItemBuildProducer) throws BuildException {
 
-        Collection<ClassInfo> allKnownImplementors = combinedIndexBuildItem.getIndex().getAllKnownImplementors(REQUEST_HANDLER);
+        List<ClassInfo> allKnownImplementors = new ArrayList<>(
+                combinedIndexBuildItem.getIndex().getAllKnownImplementors(REQUEST_HANDLER)
+                        .stream().filter(INCLUDE_HANDLER_PREDICATE).toList());
         allKnownImplementors.addAll(combinedIndexBuildItem.getIndex()
-                .getAllKnownImplementors(REQUEST_STREAM_HANDLER));
+                .getAllKnownImplementors(REQUEST_STREAM_HANDLER).stream().filter(INCLUDE_HANDLER_PREDICATE).toList());
         allKnownImplementors.addAll(combinedIndexBuildItem.getIndex()
-                .getAllKnownSubclasses(SKILL_STREAM_HANDLER));
+                .getAllKnownSubclasses(SKILL_STREAM_HANDLER).stream().filter(INCLUDE_HANDLER_PREDICATE).toList());
 
         if (allKnownImplementors.size() > 0 && providedLambda.isPresent()) {
             throw new BuildException(

--- a/extensions/amazon-lambda/deployment/src/test/java/io/quarkus/amazon/lambda/deployment/testing/LambdaWithDecoratorTest.java
+++ b/extensions/amazon-lambda/deployment/src/test/java/io/quarkus/amazon/lambda/deployment/testing/LambdaWithDecoratorTest.java
@@ -1,0 +1,83 @@
+package io.quarkus.amazon.lambda.deployment.testing;
+
+import static io.restassured.RestAssured.given;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.CoreMatchers.containsString;
+
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+
+import jakarta.annotation.Priority;
+import jakarta.decorator.Decorator;
+import jakarta.decorator.Delegate;
+import jakarta.enterprise.inject.Any;
+import jakarta.inject.Inject;
+
+import org.jboss.logging.Logger;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+
+import io.quarkus.amazon.lambda.deployment.testing.model.InputPerson;
+import io.quarkus.test.QuarkusUnitTest;
+
+class LambdaWithDecoratorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(LambdaWithDecorator.class, RequestHandlerDecorator.class, InputPerson.class))
+            .setLogRecordPredicate(record -> record.getLevel().intValue() == Level.INFO.intValue()
+                    && record.getMessage().contains("handling request with id"))
+            .assertLogRecords(records -> assertThat(records)
+                    .extracting(LogRecord::getMessage)
+                    .isNotEmpty());
+
+    @Test
+    public void testLambdaWithDecorator() throws Exception {
+        // you test your lambdas by invoking on http://localhost:8081
+        // this works in dev mode too
+
+        InputPerson in = new InputPerson("Stu");
+        given()
+                .contentType("application/json")
+                .accept("application/json")
+                .body(in)
+                .when()
+                .post()
+                .then()
+                .statusCode(200)
+                .body(containsString("Hey Stu"));
+    }
+
+    public static class LambdaWithDecorator implements RequestHandler<InputPerson, String> {
+
+        @Override
+        public String handleRequest(InputPerson input, Context context) {
+            return "Hey " + input.getName();
+        }
+    }
+
+    @Priority(10)
+    @Decorator
+    public static class RequestHandlerDecorator<I, O> implements RequestHandler<I, O> {
+
+        @Inject
+        Logger logger;
+
+        @Inject
+        @Any
+        @Delegate
+        RequestHandler<I, O> delegate;
+
+        @Override
+        public O handleRequest(I i, Context context) {
+            logger.info("handling request with id " + context.getAwsRequestId());
+            return delegate.handleRequest(i, context);
+        }
+    }
+}

